File: problem303.cc
// problem303.cc

#include <cstdio>

long unsigned next(long unsigned m) {
    if(m % 10 < 2)
        return m+1;
    else
        return 10*next(m/10);
}

int main() {
    long unsigned sum = 0;
    for(int n = 1; n <= 1000; ++n) {
        long unsigned m = 1;
        while(m < 2222222222222222222 && m % n != 0)
            m = next(m);
        if(m % n != 0) {
            printf("no solution found for n = %d.\n", n);
            return 1;
        }
        else {
            printf("f(%d) = %lu\n", n, m);
            sum += m / n;
        }
    }
    
    printf("sum = %lu\n", sum);
}

// sum[9998] = 648549117390

File: Problem303.hs
module Problem303 where

import ListPlus

next :: Integer -> Integer
next m
    | m `mod` 10 < 2 = m+1
    | otherwise      = 10 * next (m `div` 10)

f :: Integer -> Integer
f n = maybe (error "should not happen") id . find (\m -> m `mod` n == 0) $ iterate next 1

solution = sum (map (\n -> f n `div` n) [1..9998]) + 11112222222222222222 `div` 9999 + 1

main = print solution